### DISCUSSION TITLE
Connecting to windows 7 and below

### DISCUSSION THREAD

**Discussion by akallabeth** [FreeRDP/FreeRDP#8225] [General]
Due to outdated SSL ciphers and hashes used by these systems `FreeRDP` might no longer connect to these systems if compiled against system `OpenSSL`

To fix that a custom `legacy connect ` build is required:
1. You need to compile `OpenSSL 1.1.1q`
2. You need to compile `FreeRDP` against that
3. Now you can connect to the system you have in mind

**Comment by akallabeth** [comment]
I´ve created https://github.com/FreeRDP/FreeRDP/pull/8234 to update the flatpak build script. the resulting application is linked against OpenSSL 1.1.1 and is properly configured to work with older versions of windows.

It will start `wlfreerdp` or `xfreerdp` depending on where you´re running it, but you can override that with `flatpak run --command=xfreerdp com.freerdp.FreeRDP <argumenst for freerdp>`

**Reply by akallabeth** [reply]
There is now a flathub build of `xfreerdp` that ships with OpenSSL 1.1.1

**Comment by bmiklautz** [comment]
Currently a own build of openssl is **not** necessary.
# Ubuntu
On current version of Ubuntu (22.04/22.10) it's enough to start freerdp with a lowered TLS security level. Add the command line option  `/tls-seclevel:0`
# Fedora
On Fedora 36 (should also apply to 37) it's not so easy as the default crypto policy requires TLS1.2. Setting the LEGACY policy helps to connect but has the drawback that this is configured for the whole system:
```
# as root
echo LEGACY> /etc/crypto-policies/config
update-crypto-policies
```

**Comment by fgarcialainez** [comment]
@akallabeth @bmiklautz what about Linux Alpine 3.20? I have tried to setup OpenSSL 3.3 with tls-seclevel:0, and in spite that it seems to be configured properly (running a C program that calls SSL_CTX_get_security_level returns the value defined in openssl.cnf - i.e. 0), the RDP connection doesn't work properly (it returns a FreeRDP error about TLS certificate.

As far as I have seen, the only way of making legacy devices work properly (e.g. Windows 7) is to build guacd and FreeRDP using `openssl1.1-compat-dev`. See a related PR in which I am commenting for getting input about exactly the same problem that you are describing here.

https://github.com/apache/guacamole-server/commit/f906b14f9f7102dfaaf11bbd7c694b1efeacafc2#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R42

Thanks in advance.

**Reply by akallabeth** [reply]
@fgarcialainez well, you should accept the certificate? unrelated to this though.
[edit] you might check if the legacy cipher support is compiled in, if not that might also cause such stuff to fail.

**Comment by fgarcialainez** [comment]
@akallabeth it is not about accepting the certificate or not, as the issue is in the context of guacd 1.5.5 that links against FreeRDP (and we are ignoring the certificate validation from the guacamole client). Also, note that the same configuration, machines, etc work properly when building guacd and FreeRDP with OpenSSL 1.1.1 instead of OpenSSL 3.3.

**Reply by fgarcialainez** [reply]
@akallabeth have just tried the legacy provider solution (this path was unexplored in my side), but still same error. Note that I have default and legacy providers working together. See below:

```
/ $ openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.3.2
    status: active
  legacy
    name: OpenSSL Legacy Provider
    version: 3.3.2
    status: active
/ $
```

Any additional idea or hint?

PS: Please don't mark my initial comment as off-topic, as the problem I am experiencing matches exactly the issue described in this discussion.

**Reply by akallabeth** [reply]
@fgarcialainez ok, initially this sounded like an off topic bug.
if you run `xfreerdp` (3.10.x) then you should see a log message if required hash functions are not available.

you can also check if the openssl has common ciphers with the RDP server with `openssl s_client -connect SERVER01:3389 -prexit`
[edit] for windows 7 you might need to add `-tls1` or `-tls1_1` explicitly to get the certificate with `openssl`

**Reply by fgarcialainez** [reply]
@akallabeth we are using FreeRDP in guacd 1.5.5, so it is not possible to perform the first proposed test as the supported version is 2.11.7 (the latest FreeRDP 2 version).

On the other hand, I have tried to run proposed openssl commands in the legacy device and in a non-legacy device. In both of them I get a self-signed certificate error, which sounds reasonable as we are not using trusted certificates in our test environment. Note that in guacd there is a certificate verification skip setting that is used to handle this particular cases.

https://github.com/apache/guacamole-server/blob/1.5.5/src/protocols/rdp/rdp.c#L396

In our case, in up-to-date Windows devices we can connect to them without any problem using this setting (independently of having a self-signed certificate installed or not), being the issue only in Windows 7 devices and only when building guacd and FreeRDP with OpenSSL 1.1.1. If we build guacd and FreeRDP with OpenSSL 1.1.1 everything works properly.

**Reply by akallabeth** [reply]
@fgarcialainez well, I´d start `xfreerdp` directly (if it is installed on that system you can use `X11` forwarding per `SSH` to test) as `guacd` is a different software and we´d like to have a barebones test so we can narrow the source of the issue down.

**Reply by fgarcialainez** [reply]
@akallabeth yes, agree that would be great to try to isolate the problem, as it could be related to some integration with guacd. The problem is that xfreerdp is not available in the runtime image by default, as it is only using the freerdp2 libraries...

https://github.com/apache/guacamole-server/blob/1.5.5/Dockerfile#L142

Btw, have you ever checked that freerdp2 with OpenSSL 3.X works properly with legacy Windows 7 devices or older? My understanding reading the initial description of the discussion was that there were some known issue, and building with OpenSSL 1.1.1 was a must.

**Reply by akallabeth** [reply]
only issue is openssl settings and RDP that requires quite outdated digest/ciphers to work with older windows.
so, while certain properly configured new openssl works, older openssl does not require extra configuration.

**Reply by fgarcialainez** [reply]
> only issue is openssl settings and RDP that requires quite outdated digest/ciphers to work with older windows. so, while certain properly configured new openssl works, older openssl does not require extra configuration.

Yes, but having the legacy provider in place it doesn't seem to work neither, so the issue may be in some integration between FreeRDP and guacd. Other possibility is that it is an issue with different versions of OpenSSL 3 too (it may work in 3.0.X but not in 3.3.X), as I can see that your original post dates from September 2022.

I will keep you informed in the case that I find a workaround or explanation for this. Please do the same in your side if you find similar issues or complains related to this.

Thanks for your help!
